File: code/runtime/hvm/android/marshal.kt
// marshal.kt - Convert between Kotlin values and HVM terms.

package surf.clue.hvm

// Tag constants (must match hvm.c)
const val TAG_NUM: Int = 30
const val TAG_ERA: Int = 11
const val TAG_C00: Int = 13
const val TAG_C16: Int = 29

sealed class HvmValue {
    data class Num(val value: Int) : HvmValue()
    data class Str(val value: String) : HvmValue()
    data class Bool(val value: Boolean) : HvmValue()
    data class HvmList(val value: List<HvmValue>) : HvmValue()
    data class Handle(val id: Int) : HvmValue()
    data object Null : HvmValue()
}

data class NameIds(
    val boolTrue: Int,
    val boolFalse: Int,
    val listCons: Int,
    val listNil: Int,
    val stringCons: Int,
    val stringNil: Int,
)

data class MarshalContext(
    val handles: HandleTable,
    val ids: NameIds,
)

fun toTerm(input: Pair<MarshalContext, HvmValue>): Long {
    val (ctx, value) = input
    val ids = ctx.ids

    return when (value) {
        is HvmValue.Num ->
            HvmNative.hvmTermNewNum(value.value)

        is HvmValue.Str -> {
            var term = HvmNative.hvmTermNewCtr(ids.stringNil, 0, longArrayOf())
            for (ch in value.value.reversed()) {
                val chTerm = HvmNative.hvmTermNewNum(ch.code)
                val loc = HvmNative.hvmHeapAlloc(2)
                HvmNative.hvmHeapSet(loc, chTerm)
                HvmNative.hvmHeapSet(loc + 1, term)
                term = HvmNative.hvmTermNewCtr(ids.stringCons, 2, longArrayOf(chTerm, term))
            }
            term
        }

        is HvmValue.Bool -> {
            val name = if (value.value) ids.boolTrue else ids.boolFalse
            HvmNative.hvmTermNewCtr(name, 0, longArrayOf())
        }

        is HvmValue.HvmList -> {
            var term = HvmNative.hvmTermNewCtr(ids.listNil, 0, longArrayOf())
            for (item in value.value.reversed()) {
                val elem = toTerm(ctx to item)
                val loc = HvmNative.hvmHeapAlloc(2)
                HvmNative.hvmHeapSet(loc, elem)
                HvmNative.hvmHeapSet(loc + 1, term)
                term = HvmNative.hvmTermNewCtr(ids.listCons, 2, longArrayOf(elem, term))
            }
            term
        }

        is HvmValue.Handle ->
            HvmNative.hvmTermNewNum(value.id)

        is HvmValue.Null ->
            (TAG_ERA.toLong()) shl 56
    }
}

fun fromTerm(input: Pair<MarshalContext, Long>): HvmValue {
    val (ctx, term) = input
    val ids = ctx.ids
    val tag = HvmNative.hvmTermTag(term)

    if (tag == TAG_NUM) {
        return HvmValue.Num(HvmNative.hvmTermExt(term))
    }

    if (tag == TAG_ERA) {
        return HvmValue.Null
    }

    if (tag in TAG_C00..TAG_C16) {
        val name = HvmNative.hvmTermExt(term)
        val arity = tag - TAG_C00

        if (name == ids.boolTrue) return HvmValue.Bool(true)
        if (name == ids.boolFalse) return HvmValue.Bool(false)

        if (name == ids.stringCons || name == ids.stringNil) {
            return HvmValue.Str(readString(ctx to term))
        }

        if (name == ids.listCons || name == ids.listNil) {
            return HvmValue.HvmList(readList(ctx to term))
        }

        val fields = mutableListOf<HvmValue>()
        val loc = HvmNative.hvmTermVal(term)
        for (i in 0 until arity) {
            val field = HvmNative.hvmWnf(HvmNative.hvmHeapRead(loc + i))
            fields.add(fromTerm(ctx to field))
        }
        return HvmValue.HvmList(fields)
    }

    return HvmValue.Null
}

private fun readString(input: Pair<MarshalContext, Long>): String {
    val (ctx, startTerm) = input
    val ids = ctx.ids
    val chars = StringBuilder()
    var current = startTerm

    while (true) {
        val tag = HvmNative.hvmTermTag(current)
        if (tag !in TAG_C00..TAG_C16) break

        val name = HvmNative.hvmTermExt(current)
        if (name == ids.stringNil) break
        if (name != ids.stringCons) break

        val loc = HvmNative.hvmTermVal(current)
        val ch = HvmNative.hvmWnf(HvmNative.hvmHeapRead(loc))
        chars.append(HvmNative.hvmTermExt(ch).toChar())
        current = HvmNative.hvmWnf(HvmNative.hvmHeapRead(loc + 1))
    }

    return chars.toString()
}

private fun readList(input: Pair<MarshalContext, Long>): List<HvmValue> {
    val (ctx, startTerm) = input
    val ids = ctx.ids
    val items = mutableListOf<HvmValue>()
    var current = startTerm

    while (true) {
        val tag = HvmNative.hvmTermTag(current)
        if (tag !in TAG_C00..TAG_C16) break

        val name = HvmNative.hvmTermExt(current)
        if (name == ids.listNil) break
        if (name != ids.listCons) break

        val loc = HvmNative.hvmTermVal(current)
        val head = HvmNative.hvmWnf(HvmNative.hvmHeapRead(loc))
        items.add(fromTerm(ctx to head))
        current = HvmNative.hvmWnf(HvmNative.hvmHeapRead(loc + 1))
    }

    return items
}

File: code/runtime/hvm/wasm/marshal.ts
// marshal.ts - Convert between JS values and HVM terms.
//
// toTerm:   HvmValue -> bigint (HVM term)
// fromTerm: bigint (HVM term) -> HvmValue

import { NUM, ERA, C00, C16, ERA_TERM, isCtr, ctrArity } from '../base'
import type { HvmValue } from '../form'
import {
  hvmNum,
  hvmStr,
  hvmBool,
  hvmList,
  hvmRecord,
  hvmHandle,
  hvmNull,
} from '../form'
import type { HvmApi } from './bind'
import type { HandleTable } from './handle'

export type MarshalContext = {
  api: HvmApi
  handles: HandleTable

  // Well-known constructor name IDs. These are resolved at init
  // time via api.tableFind.
  ids: {
    boolTrue: number
    boolFalse: number
    listCons: number
    listNil: number
    stringCons: number
    stringNil: number
  }
}

export function toTerm(input: {
  ctx: MarshalContext
  value: HvmValue
}): bigint {
  const { ctx, value } = input
  const { api, handles, ids } = ctx

  switch (value.kind) {
    case 'num':
      return api.termNewNum(value.value)

    case 'str': {
      let term = api.termNewCtr({
        name: ids.stringNil,
        arity: 0,
        argsPtr: 0n,
      })
      for (let i = value.value.length - 1; i >= 0; i--) {
        const ch = api.termNewNum(value.value.charCodeAt(i))
        const loc = api.heapAlloc(2n)
        api.heapSet({ loc, term: ch })
        api.heapSet({ loc: loc + 1n, term })
        term = api.termNewCtr({
          name: ids.stringCons,
          arity: 2,
          argsPtr: loc,
        })
      }
      return term
    }

    case 'bool': {
      const name = value.value ? ids.boolTrue : ids.boolFalse
      return api.termNewCtr({ name, arity: 0, argsPtr: 0n })
    }

    case 'list': {
      let term = api.termNewCtr({
        name: ids.listNil,
        arity: 0,
        argsPtr: 0n,
      })
      for (let i = value.value.length - 1; i >= 0; i--) {
        const elem = toTerm({ ctx, value: value.value[i] })
        const loc = api.heapAlloc(2n)
        api.heapSet({ loc, term: elem })
        api.heapSet({ loc: loc + 1n, term })
        term = api.termNewCtr({
          name: ids.listCons,
          arity: 2,
          argsPtr: loc,
        })
      }
      return term
    }

    case 'record': {
      const arity = value.fields.length
      if (arity === 0) {
        return api.termNewCtr({ name: value.name, arity: 0, argsPtr: 0n })
      }
      const loc = api.heapAlloc(BigInt(arity))
      for (let i = 0; i < arity; i++) {
        const field = toTerm({ ctx, value: value.fields[i] })
        api.heapSet({ loc: loc + BigInt(i), term: field })
      }
      return api.termNewCtr({ name: value.name, arity, argsPtr: loc })
    }

    case 'handle': {
      return api.termNewNum(value.id)
    }

    case 'null':
      return ERA_TERM
  }
}

export function fromTerm(input: {
  ctx: MarshalContext
  term: bigint
}): HvmValue {
  const { ctx, term } = input
  const { api, ids } = ctx

  const tag = api.termTag(term)

  if (tag === NUM) {
    return hvmNum(api.termExt(term))
  }

  if (tag === ERA) {
    return hvmNull()
  }

  if (isCtr(tag)) {
    const name = api.termExt(term)
    const arity = ctrArity(tag)

    if (name === ids.boolTrue) return hvmBool(true)
    if (name === ids.boolFalse) return hvmBool(false)

    if (name === ids.stringCons) {
      return hvmStr(readString({ ctx, term }))
    }

    if (name === ids.stringNil) {
      return hvmStr('')
    }

    if (name === ids.listCons) {
      return hvmList(readList({ ctx, term }))
    }

    if (name === ids.listNil) {
      return hvmList([])
    }

    // Generic constructor: read fields into a record.
    const fields: HvmValue[] = []
    const loc = api.termVal(term)
    for (let i = 0; i < arity; i++) {
      const field = api.wnf(api.heapRead(loc + BigInt(i)))
      fields.push(fromTerm({ ctx, term: field }))
    }
    return hvmRecord({ name, fields })
  }

  return hvmNull()
}

function readString(input: {
  ctx: MarshalContext
  term: bigint
}): string {
  const { ctx } = input
  const { api, ids } = ctx
  const chars: string[] = []
  let current = input.term

  while (true) {
    const tag = api.termTag(current)
    if (!isCtr(tag)) break

    const name = api.termExt(current)
    if (name === ids.stringNil) break
    if (name !== ids.stringCons) break

    const loc = api.termVal(current)
    const ch = api.wnf(api.heapRead(loc))
    chars.push(String.fromCharCode(api.termExt(ch)))
    current = api.wnf(api.heapRead(loc + 1n))
  }

  return chars.join('')
}

function readList(input: {
  ctx: MarshalContext
  term: bigint
}): HvmValue[] {
  const { ctx } = input
  const { api, ids } = ctx
  const items: HvmValue[] = []
  let current = input.term

  while (true) {
    const tag = api.termTag(current)
    if (!isCtr(tag)) break

    const name = api.termExt(current)
    if (name === ids.listNil) break
    if (name !== ids.listCons) break

    const loc = api.termVal(current)
    const head = api.wnf(api.heapRead(loc))
    items.push(fromTerm({ ctx, term: head }))
    current = api.wnf(api.heapRead(loc + 1n))
  }

  return items
}
